[EntryExitInstrumenter] Move passes out of clang into LLVM default pipelines

pasko opened the pull request:
Move EntryExitInstrumenter(PostInlining=true) to as late as possible and
EntryExitInstrumenter(PostInlining=false) to an early pre-inlining stage
(but skip for ThinLTO post-link).

This should fix the issues reported in
https://github.com/rust-lang/rust/issues/92109 and
https://github.com/llvm/llvm-project/issues/52853. These are caused
by https://reviews.llvm.org/D97608.

pasko committed:
wip: Move instrumentation passes

This change is not ready for landing yet.

Move EntryExitInstrumenter(PostInlining=true) to as late as possible and
EntryExitInstrumenter(PostInlining=false) to an early pre-inlining stage
(but skip for ThinLTO post-link).

aeubanks commented on clang/lib/CodeGen/BackendUtil.cpp:
```diff
@@ -1016,6 +1000,11 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
     if (!IsThinLTOPostLink) {
       addSanitizers(TargetTriple, CodeGenOpts, LangOpts, PB);
       addKCFIPass(TargetTriple, LangOpts, PB);
+      PB.registerPipelineStartEPCallback(
```
we should be consistent and make both entry exit instrumenters part of the default pipelines, not a clang-specific add-on

this would probably be somewhere in `buildModuleSimplificationPipeline`, checking `Phase` (and make sure -O0 works by modifying `buildO0DefaultPipeline`)

pasko replied:
Done.

aeubanks commented on llvm/include/llvm/Passes/CodeGenPassBuilder.h:
```diff
@@ -670,9 +670,6 @@ void CodeGenPassBuilder<Derived, TargetMachineT>::addIRPasses(
     addPass(PartiallyInlineLibCallsPass());
 
-  // Instrument function entry and exit, e.g. with calls to mcount().
-  addPass(EntryExitInstrumenterPass(/*PostInlining=*/true));
```
don't touch this one, this is the WIP port of the codegen pipeline to the new pass manager (and this is actually what we want, that the codegen pipeline does this)

pasko replied:
Reverted. Thanks for sharing the context about this file!

aeubanks commented on llvm/lib/Transforms/Utils/EntryExitInstrumenter.cpp:
```diff
@@ -135,6 +138,65 @@ static bool runOnFunction(Function &F, bool PostInlining) {
+    "(post inlining)",
+    false, false)
+
+FunctionPass *llvm::createEntryExitInstrumenterPass() {
```
we don't need this one since we're adding it in the optimization pipeline which only works with the new pass manager

pasko replied:
Removed.

aeubanks commented:
can you add links to https://reviews.llvm.org/D97608, https://github.com/rust-lang/rust/issues/92109, https://github.com/llvm/llvm-project/issues/52853

pasko committed:
Address the first round of review comments

pasko commented:
> can you add links to https://reviews.llvm.org/D97608, [rust-lang/rust#92109](https://github.com/rust-lang/rust/issues/92109), #52853

Done (Updated the toplevel comment on the PR)

pasko commented:
@aeubanks Thanks for the first set of comments. I think I addressed them all and checked that instrumentation is still WAI in my ThinkLTO example. I did not run/update tests yet, this is TBD. Another round of review pre-tests would be appreciated.

aeubanks commented on llvm/include/llvm/InitializePasses.h:
```diff
@@ -101,6 +101,7 @@ void initializeEarlyMachineLICMPass(PassRegistry&);
 void initializeEarlyTailDuplicatePass(PassRegistry&);
 void initializeEdgeBundlesPass(PassRegistry&);
 void initializeEHContGuardCatchretPass(PassRegistry &);
+void initializeEntryExitInstrumenterPass(PassRegistry&);
```
not necessary

pasko replied:
Removed.

aeubanks commented on llvm/lib/Passes/PassBuilderPipelines.cpp:
```diff
@@ -1028,6 +1029,14 @@ PassBuilder::buildModuleSimplificationPipeline(OptimizationLevel Level,
+  // Instrument function entry and exit before all inlining.
+  if (Phase != ThinOrFullLTOPhase::ThinLTOPostLink &&
+      Phase != ThinOrFullLTOPhase::FullLTOPostLink &&
+      Phase != ThinOrFullLTOPhase::None) {
```
we need this pass when `Phase == ThinOrFullLTOPhase::None` right?

I'd extract out `isLTOPostLink` to mirror the existing `isLTOPreLink`

pasko replied:
I was confused by setting None prior to that in `buildPerModuleDefaultPipeline`, but it is never called for this assignment to happen. Indeed, since None means non-LTO cases, we should instrument with None. Done.

aeubanks reviewed (commented):
looks pretty good to me

for testing the pre-inliner one, we should add some tests in `llvm/test/Transforms/EntryExitInstrumenter/` that invoke things like `opt -passes='default<O1>'`, `opt -passes='thinlto-pre-link<O2>'`, `opt -passes='thinlto<O3>'` to make sure that the pass did/didn't insert the call given some IR with the appropriate function attribute

for testing the post-inliner one, an `llc` x86-64 test that checks that a call to the function was generated in the assembly given some IR with the appropriate function attribute is enough

pasko committed:
Address 2nd round of comments

pasko committed:
Run prelink instrumentation without LTO, plus tests

A few more tests to go, still a WIP.

Change a number of existing tests to swallow the newly inserted pass
name.

Add two new tests:
* CodeGen/X86/instrument-function-inlined.ll
* llvm/test/Transforms/EntryExitInstrumenter/pre-inliner-instrumentation.ll

pasko commented:
Hello @aeubanks,

I added a couple of new tests as you suggested. I do not have good intuition on their coverage. I think I will need to add a few cases, like always_inline or slightly more complicated code. Can you please take a look at the general structure? Does it roughly match your proposal above?

I would appreciate if you leave comments with specific suggestions if they come to your mind immediately.

This is still a WIP, but I thought that this PR would converge faster if I publish the current state now. A few trivial changes are still needed to tests, and I did not remove Clang tests yet.

aeubanks commented on llvm/test/Transforms/EntryExitInstrumenter/pre-inliner-instrumentation.ll:
```diff
@@ -0,0 +1,31 @@
+; RUN: opt -passes="default<O1>" -S < %s | FileCheck %s
+; RUN: opt -passes="thinlto-pre-link<O2>" -S < %s | FileCheck %s
+; RUN: opt -passes="thinlto-pre-link<O2>,thinlto<O3>" -S < %s | FileCheck %s
```
no need to run `thinlto<O3>` here, but we should test that `thinlto`/`lto` don't run this pass by adding different RUN lines with a different `FileCheck %s --checkprefix=NO`

pasko replied:
Ah, good idea! I still went with O2 for lto/thinlto because even at O1 it does not inline the alwaysinline leaf, oops.

aeubanks commented on llvm/test/Transforms/EntryExitInstrumenter/pre-inliner-instrumentation.ll:
```diff
@@ -0,0 +1,31 @@
+; RUN: opt -passes="default<O1>" -S < %s | FileCheck %s
```
should also test `O0` versions of these, plus `lto-pre-link`

pasko replied:
Done, I think, but it is evening, and I might be missing something.

aeubanks commented on llvm/test/Transforms/EntryExitInstrumenter/pre-inliner-instrumentation.ll:
```diff
@@ -0,0 +1,31 @@
+
+target triple = "x86_64-unknown-linux"
+
+define void @leaf_function() #0 {
```
mark this function as `alwaysinline` so testing `-O0` works

pasko replied:
Done.

aeubanks commented on llvm/test/CodeGen/X86/instrument-function-inlined.ll:
```diff
@@ -0,0 +1,29 @@
+define void @leaf_function() #0 {
+; CHECK-LABEL: leaf_function:
+; CHECK:       callq __cyg_profile_func_enter_bare
+; CHECK:       movq	leaf_function@GOTPCREL(%rip), %rdi
```
probably don't need this CHECK line, it just makes the test more brittle

pasko replied:
Yeah, I was also thinking that the use of a particular relocation makes it more brittle. I think it is nice to check that func_exit is used without an argument (unlike `__cyg_profile_func_enter_bare`). This can be done with the `CHECK-NEXT` on the line that follows. Though I did not notice this kind of parameter passing is already checked in `llvm/test/Transforms/EntryExitInstrumenter/mcount.ll`, hence now I am not sure how much value there is in this check.

I would not probably regret removing the line, but I'd like to try one more thing:
```
; CHECK:       {{.*}} %rdi
; CHECK-NEXT:  callq  __cyg_profile_func_exit
```

Will it blend?

pasko replied:
OK, I removed the argument passing check

pasko committed:
Address review comments for the new tests

Also tweak more pass manager tests to make them pass.

pasko committed:
Remove a couple of clang tests

pasko committed:
Add one more -O0 check in pre-inliner

pasko committed:
Revert "Remove a couple of clang tests"

This reverts commit e9c64d5c1a91e98846238c47713e90eccfe8808d.

pasko commented on llvm/test/Transforms/EntryExitInstrumenter/pre-inliner-instrumentation.ll:
```diff
@@ -0,0 +1,44 @@
+entry:
+  ret void
+; PRELTO-LABEL: entry:
+; PRELTO-NEXT:  %0 ={{( tail)?}} call ptr @llvm.returnaddress(i32 0)
```
Done.

pasko commented on llvm/test/Transforms/EntryExitInstrumenter/pre-inliner-instrumentation.ll:
```diff
@@ -0,0 +1,44 @@
+; RUN: opt -passes="default<O0>" -S < %s | FileCheck -check-prefix=PRELTO %s
```
Done. Went with INSTRUMENT/NOINSTRUMENT.

pasko commented on llvm/test/Transforms/EntryExitInstrumenter/pre-inliner-instrumentation.ll:
```diff
@@ -0,0 +1,44 @@
+; RUN: opt -passes="default<O0>" -S < %s | FileCheck -check-prefix=PRELTO %s
+; RUN: opt -passes="default<O1>" -S < %s | FileCheck -check-prefix=PRELTO %s
+; RUN: opt -passes="thinlto-pre-link<O0>,thinlto<O0>" -S < %s | FileCheck -check-prefix=PRELTO %s
```
Done.

pasko committed:
Address comments in new tests

pasko committed:
Merge branch 'main' into the instrumentation WIP PR

pasko commented on clang/test/Frontend/gnu-mcount.c:
```diff
@@ -1,41 +0,0 @@
-}
-
-// CHECK-LABEL: f
-// TODO: add profiling support for arm-baremetal
```
Done.

pasko committed:
For recovered tests in Clang only check attributes and disable llvm passes

aeubanks reviewed (commented):
forgot to hit "submit review"...

pasko committed:
Convert more Clang tests to only check attributes

aeubanks reviewed (commented):
looks pretty good, can you update the commit title to something like `[EntryExitInstrumenter] Move passes out of clang into LLVM default pipelines`

I've added test coverage for mcount-aix in 3ec57a7ed60a19c5e3e18655e19c997a469d10ec, can you merge that in?

aeubanks commented on llvm/lib/Passes/PassBuilderPipelines.cpp:
```diff
@@ -1030,6 +1036,12 @@ PassBuilder::buildModuleSimplificationPipeline(OptimizationLevel Level,
 
+  // Instrument function entry and exit before all inlining.
+  if (!isLTOPostLink(Phase)) {
+    MPM.addPass(createModuleToFunctionPassAdaptor(
```
actually is it possible to add this into `EarlyFPM` below? it's nice to have fewer module->function adaptors. (e.g. in `EarlyFPM` we run all the function passes on a single function before moving to the next one which is nice for memory locality)

sorry for suggesting this so late, since it'll require updating all the pipeline tests... but first check that none of the other tests fail

aeubanks replied:
the check for `Phase != ThinOrFullLTOPhase::FullLTOPostLink` is unnecessary, `buildModuleSimplificationPipeline` isn't called for FullLTO post link

I'm not sure why you're seeing that behavior if we're only running the post-inline instrumenter once in the codegen pipeline

aeubanks replied:
if you have a small local repro, you can pass `-Wl,-mllvm,-print-after-all` to the link command to have it print IR after every pass and see where the calls are getting added. or to see changes in the pre-link optimization pipeline, `-mllvm=-print-after-all` (or `-mllvm=-print-changed=quiet`, which doesn't work with the IR parts of the codegen pipeline for *reasons*)

pasko replied:
> the check for `Phase != ThinOrFullLTOPhase::FullLTOPostLink` is unnecessary, `buildModuleSimplificationPipeline` isn't called for FullLTO post link

Removed. Indeed there is an assertion for this a few lines prior.

> I'm not sure why you're seeing that behavior if we're only running the post-inline instrumenter once in the codegen pipeline

Apparently I made a mistake when building the reproducer. When checking today I saw everything working as expected, i.e. <=1 bare hooks inserted per function in the final DSO. Sorry for the noise.

pasko committed:
Merge branch 'main'

pasko commented on clang/test/CodeGen/mcount-aix.c:
```diff
@@ -1,25 +1,13 @@
 // RUN: %clang_cc1 -pg -triple powerpc-ibm-aix7.2.0.0 -emit-llvm %s -o - | FileCheck %s
```
Done.

pasko committed:
Run clang/test/CodeGen/... instrumentation tests only with -disable-llvm-passes

pasko committed:
Move EntryExitInstrumenterPass to EarlyFPM

pasko committed:
remove unnecessary condition

pasko committed:
Adjust passes in tests

pasko committed:
Whitespace beautify

aeubanks commented:
looks like `CodeGen/AMDGPU/llc-pipeline.ll` is failing

pasko committed:
Add the pass in llvm/test/CodeGen/AMDGPU/llc-pipeline.ll

pasko committed:
Merge 'main'

pasko commented:
> looks like `CodeGen/AMDGPU/llc-pipeline.ll` is failing

Fixed.

aeubanks commented:
thanks for doing this!
